---
title: "DO1200"
author: "Brian Yandell"
date: "2025-06-12"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

This is a reframing of DO1200 data to use with 
[qtl2shiny](https://github.com/byandell-sysgen/qtl2shiny).
See
[Recla.Rmd](Recla.Rmd)
for prototype analysis with Recla et al. dataset.
See
[DO1200Data.Rmd](DO1200Data.Rmd)
for code to set up data.

This is still very detailed. Want to decide on a few plots
and streamline code. Goal:

- scan using allele probabilities and show LOD and BLUE plot
- SNP association to find best SNP and show pattern plot
- use best SNP as covariate (not done yet)
- mediate using mRNA data (not done yet)

```{r}
(project_df <- data.frame(project = "DO1200",
                           taxa = "CCmouse",
                           directory = "qtl2shinyData"))
```

## Key Summary Information


```{r}
project <- project_df$project
project_path <- paste(rev(project_df), collapse = "/")
```

```{r}
class_name <- "plasma_metabolites"
key_trait <- "AUC_ARG"
```

Phenotype classes in current DO1200 database.

```{r}
qtl2shiny::project_classes(project_df)
```

```{r}
(peaks_df <- qtl2shiny::read_peaks(project_df, class_name) |>
  dplyr::filter(grepl(key_trait, phenotype))) |>
  dplyr::select(phenotype, qtl_lod, qtl_chr, qtl_pos)
```

```{r}
pheno_mx <- qtl2shiny:::read_pheno(project_df, class_name)
pheno_names <- colnames(pheno_mx)[grep(key_trait, colnames(pheno_mx))]
pheno_mx <- pheno_mx[, pheno_names, drop = FALSE]
pheno_mx <- apply(pheno_mx, 2, qtl2shiny:::rankZ)
colnames(pheno_mx)
```

```{r}
signif(
  data.frame(mean = apply(pheno_mx, 2, mean, na.rm = TRUE),
             sd = apply(pheno_mx, 2, sd, na.rm = TRUE)),
  4)
```

```{r}
pheno_name <- peaks_df$phenotype
chr_id <- as.character(peaks_df$qtl_chr)
peak_Mbp <- peaks_df$qtl_pos
addcovar <- peaks_df$addcovar
intcovar <- peaks_df$intcovar
window_Mbp <- 2
```

```{r}
#** Newer version of covar available, and problem with DOgen column.
#** New version has Mouse column.
covar_df <- qtl2shiny:::read_project(project_df, "covar")
covar_df$DOgen <- as.character(covar_df$DOgen)
K_chr <- qtl2shiny:::read_project(project_df, "kinship")[chr_id]
# Following used for mediation.
pmap_obj <- qtl2shiny:::read_project(project_df, "pmap")
```

## Haplotype Server

The `sex_type` was set up for DO500, which only had one diet;
for DO1200 `sex_type` = "A" corresponds to `dietset` = "all", "HC" or "HF"
in selection of `peaks` via `read_peakset()` above.

```{r}
sex_type <- "A"
```

```{r}
start_val <- peak_Mbp - window_Mbp
end_val <- peak_Mbp + window_Mbp
# Define query_probs function
query_probs <- qtl2shiny:::read_query_rds(project_df, "query_probs.rds")
# Note probs object keeps map with it
probs_obj <- query_probs(chr_id, start_val, end_val)
```

### Small region scan

This needs to be redone, walking through `qtl2mediate::scan1covar`.
Probably a new function `qtl2shiny::scan1covar`.

```{r}
dplyr::select(peaks_df, phenotype, addcovar, intcovar)
```

```{r}
scan_obj <- qtl2shiny::scan1covar(pheno_mx, covar_df, probs_obj$probs, K_chr, peaks_df)
```

```{r}
scan_window <- c(start_val, end_val)
qtl2shiny:::plot_scan(scan_obj, probs_obj$map, seq(ncol(scan_obj)), chr_id, 
                      scan_window, pheno_mx)
```

```{r}
addcovar <- qtl2shiny:::covar_model_matrix(peaks_df$addcovar, covar_df)[[1]]
allele_info <- qtl2shiny::read_project(project_df, "allele_info")
haplos <- allele_info$code
```

```{r}
blups <- FALSE
eff_obj <- qtl2ggplot::listof_scan1coef(probs_obj$probs, pheno_mx, K_chr,
                                        addcovar, blups)
qtl2shiny:::plot_eff(pheno_name, eff_obj, probs_obj$map, scan_obj, scan_window,
                 addlod = TRUE, allele_info)
```

```{r eval=FALSE}
blups <- TRUE
addcovar <- peaks_covar(peaks_df$addcovar, covar)
eff_obj <- qtl2ggplot::listof_scan1coef(probs_obj$probs, pheno_mx, K_chr,
                                        addcovar, blups)
allele_info <- qtl2shiny:::read_project(project_df, "allele_info")
qtl2shiny:::plot_eff(pheno_name, eff_obj, probs_obj$map, scan_obj, scan_window,
                 addlod = TRUE, allele_info)
```

```{r}
blups <- FALSE
eff_arg1 <- qtl2ggplot::listof_scan1coef(probs_obj$probs,
                                         pheno_mx[,1, drop=FALSE], K_chr,
                                         addcovar, blups)
qtl2shiny:::plot_eff(pheno_names[1], eff_arg1, probs_obj$map, scan_obj, scan_window,
                 addlod = TRUE, allele_info)
```

## Lipid Hotspot

Look at liver_lipids that map to the same place.
Is the signature the same?
Look at additive and diet_interactive hotspots.

```{r}
lipid_df <- qtl2shiny::read_peaks(project_df, "liver_lipids")
lipid_hotspot <- subset(qtl2shiny::hotspot(pmap_obj, lipid_df), chr = "18")
```

```{r}
lipid_diet_df <- qtl2shiny::read_peaks(project_df, "liver_lipids",
  subject_model = "all_mice_diet_interactive")
lipid_diet_hotspot <- subset(qtl2shiny::hotspot(pmap_obj, lipid_diet_df),
                             chr = "18")
```

```{r}
lipids_hot <- cbind(lipid_hotspot, lipid_diet_hotspot,
                    scannames = c("all", "lipid_add", "lipid_diet"))
```

```{r}
qtl2shiny:::plot_hot(lipids_hot, class = 2:3)
```

### Harvest Liver Lipid Phenotypes

```{r}
lipid_diet_df <- qtl2shiny:::peaks_in_pos(lipid_diet_df, TRUE,
                   chr_id, peak_Mbp, window_Mbp)
```

```{r}
dplyr::count(lipid_diet_df, addcovar, intcovar)
```

```{r}
lipid_names <- lipid_diet_df$phenotype
pheno_lipid <- qtl2shiny::read_pheno(project_df, "liver_lipids",
                                     columns = lipid_names)
```

```{r}
scan_lipid_obj <- qtl2shiny::scan1covar(pheno_lipid, covar_df, probs_obj$probs,
                                        K_chr, lipid_diet_df)
```

```{r}
saveRDS(scan_lipid_obj, "scan_lipid_obj.rds")
```

```{r}
qtl2shiny:::plot_scan(scan_lipid_obj, probs_obj$map, seq(ncol(scan_lipid_obj)),
                      chr_id, 
                      scan_window, pheno_lipid) +
  ggplot2::theme(legend.position = "none")
```

### SNP Association and SDP Plots

Note that the `query_probs()` function is set up to read genotype probabilities
for alleles and allele pairs (depending of option `allele`).
In future, this could read the `snpprobs_fstindex.rds` instead of going through
the process of recomputing from allele pairs as is done below.

Goal once we have SNP for max association, is to add the SNP genoprobs for that
SNP as additive covariate. May need to modify `scan1covar()` above to enable
that, say by adding an argument.

`snpSetupServer`

**NOTE change of snpprobs_obj below to use precomputed values. Change call below.**

```{r}
snpprobs_obj <- query_probs(chr_id, peak_Mbp - window_Mbp,
                            peak_Mbp + window_Mbp, allele = "snp")
      
      # define the query_variants function
query_variants <- qtl2shiny:::read_query_rds(project_df, "query_variants.rds")
snpinfo <- query_variants(chr_id, peak_Mbp - window_Mbp,
                          peak_Mbp + window_Mbp)
## **Need to change this**
snpprobs_obj <- qtl2mediate::get_snpprobs(chr_id, peak_Mbp, window_Mbp,
  pheno_name, probs_obj$probs, probs_obj$map, snpinfo)
```

```{r}
snp_action <- "basic"
snpprobs_act <- qtl2pattern::snpprob_collapse(snpprobs_obj$snpprobs, snp_action)
snp_scan_obj <- qtl2shiny::scan1covar(pheno_mx, covar_df, snpprobs_act, K_chr, peaks_df)
```

```{r}
minLOD <- max(3, round(max(unclass(snp_scan_obj)), 1) - 1.5)
drop_hilit <- max(unclass(snp_scan_obj)) - minLOD
top_snps_tbl <- qtl2pattern::top_snps_pattern(snp_scan_obj,
  snpprobs_obj$snpinfo, drop_hilit)
```

Show max LOD for unique consequences.

```{r}
knitr::kable(
  dplyr::ungroup(
    dplyr::filter(
      dplyr::group_by(
        dplyr::select(
          top_snps_tbl,
#         dplyr::mutate(top_snps_tbl, consequence = abbreviate(consequence, 45)),
          snp_id, pos, sdp, lod, consequence),
        consequence),
      lod == max(lod), !is.na(consequence))))
```

```{r}
(patterns <- 
  dplyr::arrange(
    dplyr::mutate(
      dplyr::filter(
        summary(top_snps_tbl), 
        .data$max_lod >= 3), 
      contrast = snp_action), 
    dplyr::desc(.data$max_lod)))
```

#### ARG1

```{r}
snp_scan_arg1 <- 
  subset(
    qtl2shiny::scan1covar(pheno_mx[,c(1,4), drop=FALSE], covar_df,
                          snpprobs_act, K_chr, peaks_df),
    lodcolumn = 2)
```

```{r}
minLOD <- max(3, round(max(unclass(snp_scan_arg1)), 1) - 1.5)
drop_hilit <- max(unclass(snp_scan_arg1)) - minLOD
top_snps_tbl1 <- qtl2pattern::top_snps_pattern(snp_scan_arg1,
  snpprobs_obj$snpinfo, drop_hilit)
```

Show max LOD for unique consequences.

```{r}
knitr::kable(
  dplyr::ungroup(
    dplyr::filter(
      dplyr::group_by(
        dplyr::select(
          top_snps_tbl1,
#         dplyr::mutate(top_snps_tbl, consequence = abbreviate(consequence, 45)),
          snp_id, pos, sdp, lod, consequence),
        consequence),
      lod == max(lod), !is.na(consequence))))
```

```{r}
patterns1 <- 
  dplyr::arrange(
    dplyr::mutate(
      dplyr::filter(
        summary(top_snps_tbl), 
        .data$max_lod >= 3), 
      contrast = snp_action), 
    dplyr::desc(.data$max_lod))
```

#### SNP Association Maps

`snpSetupServer`: `snpGeneServer`: `snpPlotServer`

```{r}
qtl2shiny:::top_snp_asso(snp_scan_obj, snpprobs_obj$snpinfo,
  scan_window, snp_action, minLOD = minLOD)
```

```{r}
qtl2shiny:::top_snp_asso(snp_scan_arg1, snpprobs_obj$snpinfo,
  scan_window, snp_action, minLOD = minLOD)
```

`snpSetupServer`: `snpPatternServer`

```{r}
dropHilit <- max(0, max(unclass(snp_scan_obj)) - minLOD)
qtl2shiny:::top_pat_plot(pheno_name, snp_scan_obj, chr_id, snpprobs_obj$snpinfo, 
  scan_window, drop_hilit = dropHilit, facet = "pattern",
  snp_action = snp_action)
```

```{r}
dropHilit <- max(0, max(unclass(snp_scan_arg1)) - minLOD)
qtl2shiny:::top_pat_plot(pheno_names[1], snp_scan_arg1, chr_id, snpprobs_obj$snpinfo, 
  scan_window, drop_hilit = dropHilit, facet = "pattern",
  snp_action = snp_action)
```

## Triad Plots of Arginine

Notice `pairprobs_obj` has 36 genotypes in second dimension of `chr_id`.

```{r}
pairprobs_obj <- query_probs(chr_id, start_val, end_val, allele = FALSE)
dim(probs_obj$probs)
dim(pairprobs_obj$probs)
```

```{r}
snpinfo <- query_variants(chr_id,
                                peak_Mbp - window_Mbp,
                                peak_Mbp + window_Mbp)
snpPairProbs_obj <- qtl2mediate::get_snpprobs(chr_id, peak_Mbp, window_Mbp,
                                pheno_name, 
                                pairprobs_obj$probs,
                                pairprobs_obj$map,
                                snpinfo)
```

```{r}
target <- pheno_mx[,4, drop = FALSE]
mediator <- pheno_mx[,-4]
# Need to get SNP that is in top_snps_tbl
snp_id <- rownames(max(snp_scan_obj, snpPairProbs_obj$snpinfo))
driver <- snpPairProbs_obj$snpprobs[[chr_id]][,,snp_id]
driver_add <- apply(driver[,c(1,3)], 1, diff) / 2
driver_dom <- driver[,2] - driver_add
```

```{r}
arg_triad <- intermediate::mediation_triad(target, mediator, driver,
                        covar_tar = addcovar, covar_med = addcovar,
                        fitFunction = qtl2mediate::fitQtl2,
                        kinship = K_chr)
ggplot2::autoplot(arg_triad)
```

```{r}
arg_triad1 <- intermediate::mediation_triad(target, mediator[,1, drop=FALSE],
                                           driver,
                        covar_tar = addcovar, covar_med = addcovar,
                        fitFunction = qtl2mediate::fitQtl2,
                        kinship = K_chr)
ggplot2::autoplot(arg_triad1)
```

```{r}
arg_triad2 <- intermediate::mediation_triad(target, mediator[,2, drop=FALSE],
                                           driver,
                        covar_tar = addcovar, covar_med = addcovar,
                        fitFunction = qtl2mediate::fitQtl2,
                        kinship = K_chr)
ggplot2::autoplot(arg_triad2)
```

```{r}
arg_triad3 <- intermediate::mediation_triad(target, mediator[,3, drop=FALSE],
                                           driver,
                        covar_tar = addcovar, covar_med = addcovar,
                        fitFunction = qtl2mediate::fitQtl2,
                        kinship = K_chr)
ggplot2::autoplot(arg_triad3)
```


## F2-Like SDP Scans

`diploServer` : `patternServer` and `alleleServer`

The panel labeled `SNP/Gene Action` has SNP association and
allele pattern that are similar to the above.
The difference is that the above uses 2-level SNPs based on the alleles
while this panel uses the 3-level SNPs.
As code is written in qtl2shiny, the SNP probabilities are derived from the
allele or allele pair genotype probabilities.
Since the new DO1200 cross has the 3-level SNPs already computed,
those could be used.

We skip that for now and consier the third action called
`Genome Scans`.
In this action, the genotype probabilities are recast according to
a selected SDP (strain distribution pattern).
That is, one can take the 36-allele pair genotype probabilities and
compute the 3-level genotype probability for the SDP of _any_ selected SNP.
This enables one to do a genome scan with a synthetic F2-like cross,
althought the probabilities are more like the SNP probabilities for the DO.
For instance the SDP `ABH:CDEFG` would count the number of `ABH` alleles
(0, 1, or 2 copies of any of `A`, `B` and/or `H`)
at each marker loci.

This is only set up for one phenotype (?) and additive covariates only.

```{r}
pats <- qtl2shiny:::pull_patterns(patterns, colnames(pheno_mx))
pattern <- pats$pattern[1]
```

```{r}
scan_pat <- qtl2pattern::scan1pattern(pairprobs_obj$probs,
                            pheno_mx,
                            K_chr, addcovar,
                            pairprobs_obj$map,
                            pats,
                            blups = blups)
```

```{r}
qtl2shiny:::scan_pat_type(scan_pat, pairprobs_obj$map, "lod", pats$pattern, 
              pheno_name, haplos)
```

```{r}
pairprobs_all <- query_probs(chr_id, allele = FALSE)
dim(pairprobs_all$probs)
```

```{r}
scan_pat_all <- qtl2pattern::scan1pattern(pairprobs_all$probs,
                            pheno_mx,
                            K_chr, addcovar,
                            pairprobs_all$map,
                            pats,
                            blups = blups)
```

```{r}
qtl2shiny:::scan_pat_type(scan_pat_all, pairprobs_all$map, "lod", pats$pattern, 
              pheno_name, haplos)
```


## Mediation

`haploServer`: `mediateServer`

```{r}
pos_Mbp <- peak_Mbp
qtls <- 1 # 1 or 2
other <- FALSE
med_plot_type <- "pos_LR"
local <- FALSE
signif <- TRUE
```

Assume comediator rather than expression.

** The `pheno_region` is not working properly. **

```{r}
comediator_region <- function(pheno_name, chr_id, scan_window, 
                              covar, peaks, 
                              qtls = 2, pmap, pheno_mx) {
  peaks_not <- dplyr::filter(peaks, .data$phenotype != pheno_name)
  # Filter peaks and analyses to region and drop pheno_name
  peaks_local <- dplyr::filter(peaks_not,
                               .data$qtl_chr == chr_id,
                               .data$qtl_pos >= scan_window[1],
                               .data$qtl_pos <= scan_window[2])
  # Read the phenos we need.
  phenos <- peaks_local$phenotype
  pheno_mx_local <- pheno_mx[, phenos, drop = FALSE]

  # Create comediator object.
  out <- pheno_region(chr_id, scan_window, covar, pmap,
    peaks, pheno_mx_local, drivers = qtls)
  
  out
}
```

** remove analyses mention **

```{r}
pheno_region <- function(chr_id, scan_window, covar, map, 
                         peaks, pheno_data,
                         drivers = 2) {
  
  start_val <- scan_window[1]
  end_val <- scan_window[2]
  # Replace NA covariate calls by FALSE.
  covars <- lapply(peaks$addcovar,peaks_covar, covar)
  # *** NEEDS WORK ***

  # Match below by pheno and other optional columns. Used in 'qtl2shiny'.
  bycols <- c("pheno", "longname", "output", "pheno_group", "pheno_type")
  m <- match(bycols, names(peaks))
  bycols <- bycols[!is.na(m)]

  ## Annotation
  annot <- 
    dplyr::rename(
      dplyr::inner_join(
        peaks, 
        dplyr::ungroup(
          dplyr::summarize(
            dplyr::group_by(
              peaks, 
              .data$phenotype),
            qtl_ct = dplyr::n(),
            info = paste0(.data$qtl_chr, "@",
                          round(.data$qtl_pos), ":",
                          round(.data$qtl_lod), collapse = ","))),
        by = "phenotype"),
      id = .data$phenotype)

  # Used in 'qtl2shiny'.
  if("pheno_type" %in% names(annot)) {
    annot <- dplyr::rename(
      annot,
      biotype = .data$phenotype)
  }
  
  # Reduce to phenotypes with peaks in region.
  annot <- dplyr::filter(
    annot, 
    .data$qtl_chr == chr_id,
    .data$qtl_pos >= start_val,
    .data$qtl_pos <= end_val)
  
  # This limits to traits that reside locally. Only make sense for expression data.
  annot$local <- FALSE
  
  # Identify markers for drivers of mediators.
  if(drivers == 2)
    annot$driver <- qtl2::find_marker(map, chr_id, annot$pos)
  
  # Make sure some pheno_data have peaks.
  m <- match(annot$id, colnames(pheno_data))
  if(any(is.na(m)))
    return(NULL)
  
  # Kludge to get names of covariates that are used by comediators.
  covars <- apply(analyses[, covars], 2, any)
  covars <- names(covars)[covars]
  covar <- covar[,covars, drop = FALSE]
  
  # Transform data if needed.
  pheno_data <- pheno_trans(
    pheno_data[, m, drop = FALSE],
    analyses$pheno, analyses$transf,
    analyses$offset, analyses$winsorize)
  
  list(pheno = pheno_data,
       annot = annot, 
       covar = covar)
}
```


```{r}
comed_ls <- comediator_region(pheno_name, chr_id, scan_window,
  covar, peaks, qtls, pmap_obj, pheno_mx)
med_ls <- qtl2mediate::comediator_type(comed_ls, peaks, pheno_name, other)
```

```{r}
phe1_mx <- if(pheno_name %in% colnames(pheno_mx)) {
    pheno_mx[, pheno_name, drop = FALSE]
  } else {
    NULL
  }
mediate_obj <- qtl2mediate::mediation_test_qtl2(
          target = phe1_mx,
          mediator = med_ls[[1]],
          annotation = med_ls[[2]],
          covar_tar = cov_df,
          covar_med = med_ls$covar,
          genoprobs = probs_obj$probs,
          map = probs_obj$map,
          chr = chr_id,
          pos = pos_Mbp,
          kinship = K_chr)
```

```{r}
ggplot2::autoplot(mediate_obj, med_plot_type,
            local_only = local, 
            significant = signif) +
            ggplot2::geom_point(size = 4)
```

`triadServer`

```{r}
sdps <- unique(dplyr::filter(patterns, .data$pheno == pheno_name)$sdp)
choices_pattern <- qtl2pattern::sdp_to_pattern(sdps, haplos)
pattern <- choices_pattern[1]
sdp <- sdps[match(pattern, choices_pattern, nomatch = 1)]
medID <- "symbol"
choices_triad <- levels(shiny::req(mediate_obj)$best$triad)
triad <- choices_triad[1] 
choices_med <- dplyr::filter(mediate_obj$best, .data$triad == triad)$id
med_name <- choices_med[1]
```

```{r}
scat_dat <- qtl2mediate::mediation_triad_qtl2(
      target = pheno_mx,
      mediator = med_ls[[1]][, med_name, drop = FALSE],
      annotation = med_ls[[2]],
      covar_tar = cov_df,
      covar_med = med_ls$covar,
      genoprobs = probs_obj$probs,
      map = probs_obj$map,
      chr = chr_id,
      pos = pos_Mbp,
      kinship = K_chr[[1]],
      sdp = sdp)
```

```{r}
peak_mar <- qtl2::find_marker(probs_obj$map, chr_id, pos_Mbp)
scat_plot <- "by_mediator"
```

** For some reason, the text is not showing on the
`qtl2shiny` rendition anymore. **

```{r}
ggplot2::autoplot(scat_dat, type = scat_plot,
             dname = peak_mar,
             mname = med_name,
             tname = colnames(pheno_mx),
             fitlines = "sdp",
             centerline = NULL)
```

```{r}
ggplot2::autoplot(scat_dat, type = scat_plot,
             dname = peak_mar,
             mname = med_name,
             tname = colnames(pheno_mx),
             fitlines = "sdp-parallel",
             centerline = NULL)
```

```{r}
ggplot2::autoplot(scat_dat, type = scat_plot,
             dname = peak_mar,
             mname = med_name,
             tname = colnames(pheno_mx),
             fitlines = "parallel",
             centerline = NULL)
```

```{r}
ggplot2::autoplot(scat_dat, type = scat_plot,
             dname = peak_mar,
             mname = med_name,
             tname = colnames(pheno_mx),
             fitlines = "driver",
             centerline = NULL)
```
