---
title: "DO1200"
author: "Brian Yandell"
date: "2025-06-12"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r}
rd_dir <- "/Volumes/mkeller3/General/main_directory/mapping_data"
```

This document concerns local data structures for the
`DO1200` data project conducted in the
[AttieLab Systems Genetics](https://github.com/AttieLab-Systems-Genetics)
organization.

## DO1200 Managed Data and Code

The data are maintained on the 
[UW-Madison ResearchDrive](https://it.wisc.edu/services/researchdrive/)
in the `mkeller3` partition in `General/main_directory`,
primarily in subfolders

- `mapping_data/` data objects used for analysis (mapping)
- `files_for_cross_object/` data used to construct cross object
- `annotated_peak_summaries/` peak summaries by subject set and covariate model

The following code would load the `cross_obj` (2.7Gb) and
one of the peaks dataframe files (`peak_clinical_df`)
and one of the phenotype files.
Note that the peaks and phenotype file names have extra information
about project and versions.

```{r eval=FALSE}
cross_obj <- readRDS(file.path(rd_dir, "cross_DO_diet_grcm39_v6.rds"))
class <- "clinical_traits"
peak_clinical_df <- read.csv(file.path(rd_dir, "..", "annotated_peak_summaries",
  paste0("DO1200_", class, "_all_mice_additive_peaks.csv")))
pheno_clinical_df <- read.csv(file.path(rd_dir, "..", "files_for_cross_object",
  paste0("pheno_", "in_vivo_", class, "_v3.csv")))
```

The data are maintained with master QTL mapping scripts on GitHub at
[AttieLab-Systems-Genetics/qtl_mapping/scripts/latest](https://github.com/AttieLab-Systems-Genetics/qtl_mapping/tree/main/scripts/latest).
Note in particular the scan scripts

- [QTL_scan_wrapper.R](https://github.com/AttieLab-Systems-Genetics/qtl_mapping/blob/main/scripts/latest/QTL_scan_wrapper.R)
- [QTL_scan_function.R](https://github.com/AttieLab-Systems-Genetics/qtl_mapping/blob/main/scripts/latest/QTL_scan_function.R)

The wrapper sources the script
[rankz.R](https://github.com/AttieLab-Systems-Genetics/qtl_mapping/blob/main/scripts/latest/rankz.R)
located in the same folder for the data transformation.

## DO1200 Preparation for qtl2shiny

The `DO1200` data are prepared for use with the
[qtl2shiny](https://github.com/byandell-sysgen/qtl2shiny)
app.
The intent is to largely maintain data in the structure as the above
managed data, with adjustments for rapid access and use.
Some processing is done, for instance to put data in RDS files
for quicker access and to combine disparate phenotype files.

CSV files were transferred by hand, read and saved as RDS.
For instance, peaks came from
`file.path(rd_dir, "..", "annotated_peak_summaries")`
as CSV files, and landed in 
`qtl2shinyData/CCmouse/DO1200/peaks`
as RDS files.

Key sections:

- [Project and CC Mouse Data](#project-and-cc-mouse-data)
  - [Miscellaneous Items from Cross Object](#miscellaneous-items-from-cross-object)
- [DO1200 Genotype Data](#do1200-genotype-data)
  - [Kinship](#kinship)
  - [Genes and SNP Variants](#genes-and-snp-variants)
  - [Genotype Probabilities](#genotype-probabilities)
- [DO1200 Peak Data](#do1200-peak-data)
- [DO1200 Phenotype Data](#do1200-phenotype-data)
  - [DO1200 Covariate Data](#do1200-covariate-data)
  - [DO1200 Phenotypes by Class](#do1200-phenotypes-by-class)
  
<hr>

## Project and CC Mouse Data

The project is identified at the top level in a CSV file
`projects.csv`

```
project,taxa,directory
Recla,CCmouse,qtl2shinyData
DO1200,CCmouse,qtl2shinyData
```

Each row corresponds to a project.
The `project` is located within a `taxa`, which is located within the
overall `qtl2shinyData` folder.
For ease, we define the project info below.

```{r}
(project_df <- data.frame(project = "DO1200",
                           taxa = "CCmouse",
                           directory = "qtl2shinyData"))
```

```{r}
taxa_path <- file.path(project_df$directory, project_df$taxa)
dir(taxa_path)
```

```{r}
(project_path <- file.path(taxa_path, project_df$project))
if(!dir.exists(project_path)) dir.create(project_path, recursive = TRUE)
```

```{r}
dir(project_path)
```

### Miscellaneous Items from Cross Object

The cross object is huge and takes a long time to load. 
Most of its elements can be copied from the ResearchDrive.
However, it was convenient to (one time) save the various
miscellaneous objects into a folder.

```{r eval=FALSE}
if(!file.exists(file.path(project_path, "covar.rds"))) {
  cross <- readRDS(file.path(rd_dir, "cross_DO_diet_grcm39_v6.rds"))
  cross$pheno <- NULL
  cross$geno <- NULL
  # Pull out individual objects.
  for(x in names(cross)) {
    savepath <- file.path(project_path, "misc", paste0(x, ".rds"))
    print(savepath)
    if(!file.exists(savepath)) {
      saveRDS(cross[[x]], savepath)
    }
  }
}
```

```{r eval=FALSE}
sum_na <- function(x) sum(!is.na(x))
# Generation (or wave) of each mouse.
table(qtl2shiny::read_project(project_df, "cross_info", "misc"))

# Gene information; includes mus-human comparisons.
gene_annos <- qtl2shiny::read_project(project_df, "gene_annos", "misc")
is_human <- grep("hum", names(gene_annos))
sapply(gene_annos[,-is_human], sum_na)
sapply(gene_annos[is_human], sum_na)

# Isoform transcript information; includes mus-human comparisons.
transcript_annos <- qtl2shiny::read_project(project_df, "transcript_annos", "misc")
is_human <- grep("hum", names(transcript_annos))
sapply(transcript_annos[,-is_human], sum_na)
sapply(transcript_annos[is_human], sum_na)

# Genotype of founder across all markers (1,2,3)
sapply(qtl2shiny::read_project(project_df, "founder_geno", "misc"), dim)

rm(sum_na, gene_annos, transcript_annos, is_human)
```

## DO1200 Genotype Data

### Kinship

Kinship file `kinship_loco` is renamed `kinship`.

```{r eval=FALSE}
if(!file.exists(kinship_rds <- file.path(project_path, "kinship.rds"))) {
  kinship <- readRDS(file.path(rd_dir, "k_loco_DO1200_grcm39.rds"))
  saveRDS(kinship, kinship_rds)
}
```

```{r}
sapply(qtl2shiny::read_project(project_df, "kinship"), dim)
```

### Genes and SNP Variants

**This is not quite right. Need to know where these came from (`from_path`).
Copying these files takes a long time.

```{r eval=FALSE}
sql_files <- paste0(c("mouse_genes", "cc_variants"), "_grcm39.sqlite")
if(!all(file.exists(file.path(taxa_path, sql_files)))) {
  for(i in sql_files) {
    message("copying ", i)
    system(paste0("cp ", file.path(from_path, i), " ", taxa_path))
  }
}
```

Need to construct `query_variants.rds` and `query_genes.rds`.
Look on qtl2shiny vignette
[qtl2shinyData.Rmd](https://github.com/byandell-sysgen/qtl2shiny/blob/master/vignettes/qtl2shinyData.Rmd).
** Need to modify code to distinguish from old `gmc38` versions.**

```{r}
if(file.exists(qname <- file.path(taxa_path, "query_genes_grcm39.rds"))) {
  warning(paste("gene query", qname,
                "already exists and will not be overwritten"))
} else {
  query_genes <- 
    qtl2::create_gene_query_func(
      file.path(taxa_path, "mouse_genes_grcm39.sqlite"))
  saveRDS(query_genes, qname)
}
```

```{r}
if(file.exists(qname <- file.path(taxa_path, "query_variants_grcm39.rds"))) {
  warning(paste("variant query", qname,
                "already exists and will not be overwritten"))
} else {
  query_variants <- 
    qtl2::create_variant_query_func(
      file.path(taxa_path, "cc_variants_grcm39.sqlite"))
  saveRDS(query_variants, qname)
}
```

### Genotype Probabilities

Separate FST collections for alleles, allele pairs and SNPs.
These take a long time to copy.

```
probs_fst/
```

```{r}
list.files(file.path(rd_dir, "probs_fst"), pattern = "*.rds")
```

```{r}
if(!dir.exists(file.path(project_path, "probs_fst"))) {
  system(paste0("cp -r ", file.path(rd_dir, "probs_fst"), " ", project_path))
}
```

```{r}
dim(readRDS(file.path(project_path, "genoprob", "alleleprobs_fstindex.rds")))
```

```{r}
dim(readRDS(file.path(project_path, "genoprob", "genoprobs_fstindex.rds")))
```

```{r}
dim(readRDS(file.path(project_path, "genoprob", "snpprobs_fstindex.rds")))
```

Use `create_probs_query_func` from package 'qtl2pattern' to query genotype probabilities.
Save it in RDS format for later use. Note that it requires a relative address to the project data provided by `project_path`.

```{r}
if(file.exists(qname <- file.path(project_path, "query_probs.rds"))) {
  warning(paste("probs query", qname,
                "already exists and will not be overwritten"))
} else {
  query_probs <- qtl2pattern::create_probs_query_func(
    project_path, probdir = "probs_fst")
  saveRDS(query_probs, qname)
}
```

## DO1200 Phenotype Data

### DO1200 Covariate Data

Covariates were already retrieved from the `cross` object.
They are also available as CSV from the `files_for_cross_object` folder.
The `covar_df` object is a dataframe with mouse ID as `rownames`.

```{r eval=FALSE}
if(!file.exists(covar_rds <- file.path(project_path, "covar.rds"))) {
  covar_df <- readRDS(file.path(rd_dir, "..", "files_for_cross_object",
                                "covar.csv"))
  saveRDS(covar_df, covar_rds)
}
```

### DO1200 Phenotypes by Class

```
DO1200/pheno/
```

Phenotypes are organized by classes.
The raw data come from

```{r}
dir(file.path(rd_dir, "../files_for_cross_object"), pattern = "^pheno_.*")
```

The data were copied over as CSV, read, and saved as RDS.
These files have Mouse as first column, so are data frames.
They are saved in cross object as matrix with rownames being Mouse ID.

Right now they are saved as `pheno_<class>.rds` as dataframe,
but might be better as `pheno/<class>.rds` as matrix with rownames,
since that is how they will be used.

Note that translation from CSV to RDS can introduce artefacts,
such as `"#DIV/0!"` errors.

### Classes and Phenotypes

Phenotypes depend on `dataset`.

```{r}
classes <- dir(file.path(rd_dir, "..", "files_for_cross_object"),
                pattern = "pheno_.*")
class_names <- stringr::str_replace(
  stringr::str_remove(
    stringr::str_remove(
      stringr::str_remove(
        stringr::str_remove(
          stringr::str_remove(classes, ".csv$"),
          "_v[0-9]+$"),
        "_(DO_diet_batch|DO_diet|combat)_(corrected|nonzero150_qqnorm|nonzero150_vsd)"),
      "^pheno_"),
    "(in_vivo_|_GTT_metatraits)"),
  "((13C|2H))_metabolite", "metabolites_\\1")
names(classes) <- class_names
classes
```

The following saves `classes` as `pheno_xxx.rds` with `xxx` the
`class_names`.
This is OK for now, but may want to save as FST later.


```{r}
if(!file.exists(file.path(project_path, "pheno", "pheno_clinical_traits.rds"))) {
  for(i in names(classes)) {
    savepath <- file.path(project_path, "pheno", paste0("pheno_", i, ".rds"))
    print(savepath)
    pheno_path <- file.path(rd_dir, "..", "files_for_cross_object", classes[i])
    saveRDS(read.csv(pheno_path), savepath)
  }
}
```

#### Modify Phenotypes to make Matrices with Class Names

The `classes` above are not quite right.
The `plasma_metabolites` need to be combined, and the `liver_psi`
needs to be renamed `liver_splice_juncs`.

```{r eval=FALSE}
if(!file.exists(file.path(project_path, "pheno", "pheno_liver_splice_juncs.rds"))) {
  system(paste0("mv ",
                file.path(project_path, "pheno", "pheno_liver_psi.rds"),
                " ",
                file.path(project_path, "pheno", "pheno_liver_splice_juncs.rds")))
}
```

```{r eval=FALSE}
pm <- file.path(project_path, "pheno", "pheno_plasma_metabolites")
if(!file.exists(paste0(pm, ".rds"))) {
  pm1 <- readRDS(paste0(pm, "_13C.rds"))
  pm2 <- readRDS(paste0(pm, "_2H.rds"))
  pmall <- dplyr::full_join(pm1, pm2, by = "Mouse")
  saveRDS(pmall, paste0(pm, ".rds"))
  system(paste("rm ", paste0(pm, "_13C.rds "), paste0(pm, "_2H.rds")))
  rm(pm1, pm2, pmall)
}
```

```{r}
classes <- dir(file.path(project_path, "pheno"), pattern = "pheno_.*")
class_names <- stringr::str_remove(
  stringr::str_remove(classes, "^pheno_"),
  ".rds$")
names(classes) <- class_names
classes
```

Now, phenotype objects will be changed to have `rownames` as
the first column and saved as matrices.

```{r eval=FALSE}
phenos <- dir(file.path(project_path, "pheno"), pattern = "pheno_.*")
for(i in phenos) {
  print(i)
  phenoi <- readRDS(file.path(project_path, "pheno", i))
  if(!is.matrix(phenoi)) {
    phenoi <- tibble::column_to_rownames(phenoi, var = names(phenoi)[1])
    types <- sapply(phenoi, class)
    if(!all(types == "numeric")) {
      for(j in which(types != "numeric"))
        phenoi[[j]] <- as.numeric(phenoi[[j]])
    }
    phenoi <- as.matrix(phenoi)
    saveRDS(phenoi, file.path(project_path, "pheno", i))
  }
}
```


### Peaks by Dataset and Covariates by Phenotype

There are many peaks files.
Need to decide if plan is to combine peaks files or have a bunch of
different ones. If combined, need columns for dataset, intset and dietset.

```
DO1200/peaks.rds
```

```{r}
peaksets <- dir(file.path(rd_dir, "..", "annotated_peak_summaries"),
    pattern = "^.*_all_.*additive_peaks.csv")
peakset_names <- stringr::str_replace(
  stringr::str_remove(
    stringr::str_remove(peaksets, "^DO1200_"),
    "_all_mice_additive_peaks.csv$"),
  "splice_juncs", "psi")
names(peaksets) <- peakset_names
peaksets
```

```{r}
peak_path <- file.path(project_path, "peaks")
for(i in tools::file_path_sans_ext(dir(peak_path))) {
  print(i)
  saveRDS(read.csv(file.path(peak_path, paste0(i, ".csv"))),
          file.path(peak_path, paste0(i, ".rds")))
}
```

```{r}
dir(file.path(rd_dir, "..", "annotated_peak_summaries"),
    pattern = "^.*_clinical_traits_.*additive_peaks.csv")
```

```{r}
dir(file.path(rd_dir, "..", "annotated_peak_summaries"),
    pattern = "^.*_clinical_traits_.*_peaks.csv")
```

```{r}
# "clinical_traits", liver_genes"m "liver_isoforms", "liver_lipids", "liver_splice_juncs"
dataset <- "clinical_traits"
# "all", "HC", "HF"
dietset <- "all"
# "additive", "interactive"
intset <- "additive"
peakpath <- file.path(rd_dir, "..", "annotated_peak_summaries",
                      paste0("DO1200_", dataset, "_", dietset,
                             "_mice_", intset, "_peaks.csv"))
print(peakpath)
savepath <- file.path(project_path, "peaks.rds")
print(savepath)
saveRDS(read.csv(peakpath), savepath)
```

Hotspots object is created with `hotspot`.
** Need to walk through this to find name changes in new data.**

```{r, eval=FALSE}
hots <- qtl2shiny::hotspot(pmap_obj,
                           dplyr::rename(peaks, lod = "qtl_lod"))
saveRDS(hots, file.path(project_path, "hotspot.rds"))
```

### Analyses Table

```
DO1200/analyses.rds.  # see comments below
```

The peaks files (previous section) include model information on
`addcovar` and `intcovar`.
It would be useful to switch to this.
Most of the rest of this is dated, I think.

Here is what seems to be needed in `analyses_tbl`:

- `qtl2shiny::setupServer()`
  - `pheno_group`: general `dataset` name
  - `pheno_type`: specific `dataset` as subset of `pheno_group`, such as `Islet.mRNA.19`
  - `qtl2shiny:::set_analyses(input$dataset, input$pheno_group, analyses_tbl())`
  - `qtl2shiny:::num_pheno(character(), analyses_tbl())`
  - `dplyr::count(analyses_tbl, pheno_group, pheno_type)` (for project "AttieDO")
- `qtl2shiny::hotspotServer()`
  - `pheno_group`: one of `sort(unique(analyses_tbl$pheno_group))`
  - `dataset`: `c("all", sort(unique(analyses_tbl$pheno_type)))`
- `qtl2mediate::scan1covar()` function uses:
  - `analyses_df$model`
  - `which_covar(analyses_df)`
  - `covarset <- apply(analyses_df, 1, function(x) paste(1 * x, collapse = ""))`
  - `scanfn(probs_obj, phe_mx, kinship, cov_df, analyses_df, wh, models, ...)`
  - `which_covar` and `scanfn` are local functions

The following need to be updated.
Phenotypes are easy, as we draw from the appropriate dataset.
Covariates--can use the whole data frame `covar`.

```{r}
pheno_data <- qtl2shiny:::pheno_read(project_df, analyses_tbl)
phe_mx <- qtl2shiny:::pheno_read(project_df, analyses_df)
cov_df <- qtl2mediate::get_covar(covar, analyses_df)
```


```{r}
scan_obj <- qtl2mediate::scan1covar(phe_mx, cov_df, probs_obj$probs, K_chr,
                                    analyses_df, sex_type = sex_type)
snp_scan_obj <- qtl2mediate::scan1covar(phe_mx, cov_df, snpprobs_act, K_chr,
                                analyses_df, sex_type = sex_type)
```

### SQL for genes and variants

```
CCmouse/mouse_genes_grcm39.sqlite
CCmouse/cc_variants_grcm39.sqlite
```

```{r, eval=FALSE}
for(i in c("mouse_genes", "cc_variants"))
system(paste0("cp ",
              file.path(rd_dir, paste0(i, "_grcm39.sqlite")),
              " ",
              taxa_dir))
```

Need to construct `query_variants.rds` and `query_genes.rds`.
Look on qtl2shiny vignette
[qtl2shinyData.Rmd](https://github.com/byandell-sysgen/qtl2shiny/blob/master/vignettes/qtl2shinyData.Rmd).

```{r}
if(file.exists(qname <- file.path(taxa_dir, "query_genes.rds"))) {
  warning(paste("gene query", qname,
                "already exists and will not be overwritten"))
} else {
  query_genes <- 
    qtl2::create_gene_query_func(
      file.path(taxa_path, "mouse_genes_grcm39.sqlite"))
  saveRDS(query_genes, qname)
}
```

```{r}
if(file.exists(qname <- file.path(taxa_dir, "query_variants.rds"))) {
  warning(paste("variant query", qname,
                "already exists and will not be overwritten"))
} else {
  query_variants <- 
    qtl2::create_variant_query_func(
      file.path(taxa_path, "cc_variants_grcm39.sqlite"))
  saveRDS(query_variants, qname)
}
```

### K `loco` matrix

```
DO1200/kinship.rds from k_loco_DO_diet_grcm39.rds
```

```{r, eval=FALSE}
system(paste0("cp ",
              file.path(rd_dir, "k_loco_DO_diet_grcm39.rds"),
              " ",
              file.path(project_path, "kinship.rds")))
```

```{r}
sapply(readRDS(file.path(rd_dir, "k_loco_DO_diet_grcm39.rds")), dim)
```

```{r}
sapply(readRDS(file.path(project_path, "kinship.rds")), dim)
```

### Maps and Covariates

```
DO1200/pmap.rds
DO1200/gmap.rds
DO1200/covar.rds
```

```{r, eval=FALSE}
# Likely need to adapt to other covars as well?
# Gen = DOgen in code (DOwave also present)
# Check other covariates
save_crosser("covar")
```

```
> names(covar) # AttieDOv2
 [1] "mouse"              "sex"                "sac_date"           "partial_inflation" 
 [5] "coat_color"         "oGTT_date"          "FAD_NAD_paired"     "FAD_NAD_filter_set"
 [9] "crumblers"          "birthdate"          "diet_days"          "num_islets"        
[13] "DOgen"              "DOwave"
```

```
> names(cross$covar) # DO1200
 [1] "Gen"                    "GenLit"                 "Sex"                   
 [4] "Diet"                   "liver_lipids_batch"     "chol_batch"            
 [7] "true_TG_batch"          "KB_batch"               "plasma_13C_metab_batch"
[10] "plasma_2H_metab_batch"   
```

### Phenotypes

```
DO1200/pheno_data.rds
```

This will take the most work to redesign.

#### RNASeq

Need to review this carefully

```{r}

```

## Genoprobs for alleles, allele pairs and SNPs

```
probs_fst/
```

```{r}
list.files(file.path(rd_dir, "probs_fst"), pattern = "*.rds")
```

```{r}
dim(readRDS(file.path(rd_dir, "probs_fst", "alleleprobs_fstindex.rds")))
```

```{r}
dim(readRDS(file.path(rd_dir, "probs_fst", "genoprobs_fstindex.rds")))
```

```{r}
dim(readRDS(file.path(rd_dir, "probs_fst", "snpprobs_fstindex.rds")))
```
